src/Microsoft.Azure.WebJobs.Host/Bindings/PatternMatcher.cs

```diff
@@ -0,0 +1,263 @@
+                    candidate.DeclaringType.Assembly == method.DeclaringType.Assembly)
+                {
+                    return candidate;
+                }
+            }
+            return null;
+        }
+
+        // Name can only map to a single type. If try to map to difference types, then it's a failed match. 
+        private static bool AddGenericArg(Dictionary<string, Type> genericArgs, string name, Type type)
+        {
+            Type typeExisting;
+            if (genericArgs.TryGetValue(name, out typeExisting))
+            {
+                return typeExisting == type;
+            }
+            genericArgs[name] = type;
+            return true;
+        }
+
+        // Check if specificType is a valid instance of openType. 
+        // Return true if the types are compatible. 
+        // If openType has generic args, then add a [Name,Type] entry to the genericArgs dictionary. 
+        private static bool CheckArg(Type openType, Type specificType, Dictionary<string, Type> genericArgs)
+        {
+            if (openType == specificType)
+            {
+                return true;
+            }
+            // Is it a generic match?
+            // T, string
+            if (openType.IsGenericParameter)
+            {
+                string name = openType.Name;
+                return AddGenericArg(genericArgs, name, specificType);
+            }
+
+            // IFoo<T>, IFoo<string> 
+            if (openType.IsGenericType)
+            {
+                if (specificType.GetGenericTypeDefinition() != openType.GetGenericTypeDefinition())
+                {
+                    return false;
+                }
+
+                var typeArgs = openType.GetGenericArguments();
+                var specificTypeArgs = specificType.GetGenericArguments();
+
+                int len = typeArgs.Length;
+
+                for (int i = 0; i < len; i++)
+                {
+                    if (!AddGenericArg(genericArgs, typeArgs[i].Name, specificTypeArgs[i]))
+                    {
+                        return false;
+                    }
+                }
+                return true;
+            }
+
+            return false;
+        }
+
+        // Helper for invoking a methodInfo that returns Task<T>
+        private class GetAsyncConverterFunc<T>
+        {
+            private static Func<object, object> Work(object instance, MethodInfo method)
+            {
+                Func<object, object> converter = (input) =>
+                {
+                    Task<T> resultTask = Task.Run(() => (Task<T>)method.Invoke(instance, new object[] { input }));
+
+                    T result = resultTask.GetAwaiter().GetResult();
+                    return result;
+                };
+                return converter;
+            }
+        }
+
+        // Wrapper for matching against a static type. 
+        private class CreateViaType : PatternMatcher
```

Comment:
I think these types need better names. How about TypeMatcher, InstanceMatcher (below)? Basically nouns instead of verbs.